File: internal/document/document.go
package document

import (
	"crypto/rand"
	"errors"
	"fmt"

	"github.com/john6604/yata-collaborative-editor/internal/identifier"
	"github.com/john6604/yata-collaborative-editor/internal/protocol"
)

// Package variables for Start and End nodes
var startID = identifier.NewID("START", -1)
var endID = identifier.NewID("END", -2)

// Definition of structure
type Document struct {
	Start            *Element
	End              *Element
	ElementsByID     map[identifier.ID]*Element
	CharacterCounter int
	ClientID         string
	Clock            int
	PendingInserts   map[identifier.ID]*PendingElement
	PendingDeletes   map[identifier.ID]*PendingElement
	InsertLog        map[identifier.ID]*protocol.InsertOperation
	DeleteLog        map[identifier.ID]*protocol.DeleteOperation
}

// Function to generate a new document
func NewDocument() *Document {

	//Document attributes defined
	document := Document{CharacterCounter: 0}
	start := NewElement(*startID, nil, nil, nil, '\x00')
	end := NewElement(*endID, nil, nil, nil, '\x00')
	document.ElementsByID = make(map[identifier.ID]*Element)
	document.PendingInserts = make(map[identifier.ID]*PendingElement)
	document.PendingDeletes = make(map[identifier.ID]*PendingElement)
	document.InsertLog = make(map[identifier.ID]*protocol.InsertOperation)
	document.DeleteLog = make(map[identifier.ID]*protocol.DeleteOperation)
	document.CharacterCounter = 0
	document.ClientID = generateUUID()
	document.Clock = 0

	//Initial logic defined
	start.Right = end
	end.Left = start
	document.Start = start
	document.End = end

	//Initialize linked list
	document.ElementsByID[start.ElementID] = start
	document.ElementsByID[end.ElementID] = end

	return &document
}

// Function to generate a random UUID
func generateUUID() string {
	b := make([]byte, 16)
	_, err := rand.Read(b)
	if err != nil {
		return err.Error()
	}

	uuid := fmt.Sprintf("%x-%x-%x-%x-%x",
		b[0:4], b[4:6], b[6:8], b[8:10], b[10:])
	return uuid
}

// Function to generate an ID for an element
func (d *Document) generateElementID() *identifier.ID {
	id := identifier.NewID(d.ClientID, d.Clock)
	d.Clock++
	return id
}

// Function to insert locally in the document
func (d *Document) InsertElement(index int, character rune) (error, identifier.ID) {

	previousElement, nextElement, err := d.findVisiblePosition(index)

	if err != nil {
		return err, identifier.ID{}
	}

	id := d.generateElementID()
	insertedElement := NewElement(*id, previousElement, previousElement, nextElement, character)
	previousElement.Right = insertedElement
	nextElement.Left = insertedElement
	d.ElementsByID[insertedElement.ElementID] = insertedElement
	insertOperation := protocol.NewInsertOperation(insertedElement.ElementID, previousElement.ElementID, nextElement.ElementID, character)
	d.InsertLog[insertedElement.ElementID] = insertOperation

	d.CharacterCounter++

	return nil, *id
}

func (d *Document) integrateInsert(newID identifier.ID, originID identifier.ID, rightID identifier.ID, content rune) {

	left, right := d.findInsetionPoint(originID, rightID, newID)
	origin := d.ElementsByID[originID]

	element := NewElement(newID, origin, left, right, content)

	left.Right = element
	right.Left = element

	d.ElementsByID[newID] = element
	insertOperation := protocol.NewInsertOperation(newID, originID, rightID, content)
	d.InsertLog[newID] = insertOperation

	d.CharacterCounter++

}

// Function to insert remotely/concurrently in the document
func (d *Document) RemoteInsert(newID identifier.ID, originID identifier.ID, rightID identifier.ID, content rune) error {

	if d.ElementsByID[newID] != nil {
		return errors.New("The value was already inserted.")
	}

	if d.ElementsByID[originID] == nil || d.ElementsByID[rightID] == nil {
		d.PendingInserts[newID] = NewPending(newID, originID, rightID, content)
		insertOperation := protocol.NewInsertOperation(newID, originID, rightID, content)
		d.InsertLog[newID] = insertOperation
		return errors.New("Pending value.")
	}

	d.integrateInsert(newID, originID, rightID, content)

	d.processPending()

	d.processPendingDeletes()

	return nil
}

// Function to delete an element logically
func (d *Document) Delete(index int) (error, identifier.ID) {

	element, err := d.findVisibleElement(index)

	if err != nil {
		return err, identifier.ID{}
	}

	element.IsDeleted = true
	d.CharacterCounter--
	deleteOperation := protocol.NewDeleteOperation(element.ElementID)
	d.DeleteLog[element.ElementID] = deleteOperation

	return nil, element.ElementID
}

func (d *Document) integrateDeletion(elementID identifier.ID) {

	element := d.ElementsByID[elementID]
	element.IsDeleted = true

	d.CharacterCounter--

	deleteOperation := protocol.NewDeleteOperation(element.ElementID)
	d.DeleteLog[element.ElementID] = deleteOperation
}

// Function to delete an element remotely
func (d *Document) RemoteDelete(elementID identifier.ID) error {

	if d.ElementsByID[elementID] == nil {
		d.PendingDeletes[elementID] = NewPending(elementID, identifier.ID{}, identifier.ID{}, '\x00')
		deleteOperation := protocol.NewDeleteOperation(elementID)
		d.DeleteLog[elementID] = deleteOperation
		return nil
	}

	if d.ElementsByID[elementID].IsDeleted == true {
		return nil
	}

	d.integrateDeletion(elementID)

	return nil
}

func (d *Document) IntegrateDelta(delta protocol.Delta) error {

	var inserts []*protocol.InsertOperation
	inserts = delta.Inserts

	for _, v := range inserts {
		err := d.RemoteInsert(v.NewID, v.OriginID, v.RightID, v.Content)
		if err != nil {
			if err.Error() != "Pending value." && err.Error() != "The value was already inserted." {
				return err
			}
		}
	}

	var deletes []*protocol.DeleteOperation
	deletes = delta.Deletes

	for _, v := range deletes {
		err := d.RemoteDelete(v.TargetID)
		if err != nil {
			return err
		}
	}

	return nil
}

File: frontend/src/crdt/document.js
import { Element } from "./element"
import { ID, startNode, endNode } from "./identifier"
import { InsertOperation, DeleteOperation } from "./operations";
import { PendingElement } from "./buffer";

export class Document {
    constructor() {
        const startDocument = new Element(startNode, null, null, null, '\x00');
        const endDocument = new Element(endNode, null, null, null, '\x00');
        this.elementsByID = new Map();
        this.characterCounter = 0;
        this.clientID = crypto.randomUUID();
        this.clock = 0;
        this.pendingInserts = new Map();
        this.pendingDeletes = new Map();
        this.insertLog = new Map();
        this.deleteLog = new Map();

        // Initial Logic 
        startDocument.right = endDocument;
        endDocument.left = startDocument;
        this.start = startDocument;
        this.end = endDocument;

        // Initial links
        this.elementsByID.set(startDocument.elementID.toKey(), startDocument);
        this.elementsByID.set(endDocument.elementID.toKey(), endDocument);
    }

    generateID() {
        const id = new ID(this.clientID, this.clock);
        this.clock++;
        return id;
    }

    findVisiblePosition(index) {
        if (index > this.characterCounter || index < 0) {
            throw new Error("Inexisting position to insert character."); 
        }

        let current = this.start.right;
        let visibleIndex = 0;

        while (current !== null) {
            if (!current.isDeleted) {
                if (visibleIndex === index) {
                    return [current.left, current];
                }
                visibleIndex++;
            }
            current = current.right
        }

        return [this.end.left, this.end]
    }

    findVisibleElement(index) {
        if (index >= this.characterCounter || index < 0) {
            throw new Error("The character to delete does not exist.");
        }

        let current = this.start.right;
        let visiblePosition = 0;

        while (current !== null) {
            if (!current.isDeleted) {
                if (visiblePosition === index) {
                    return current;
                }
                visiblePosition++;
            }
            current = current.right;
        }

        throw new Error("An unexpected error has occured.");
    }

    insertElement(index, character) {
        
        if (Array.from(character).length !== 1) {
            throw new Error("Must be exactly one character.");
        }

        const [previousElement, nextElement] = this.findVisiblePosition(index);

        const id = this.generateID();
        const insertedElement = new Element(id, previousElement, previousElement, nextElement, character);
        previousElement.right = insertedElement;
        nextElement.left = insertedElement;
        this.elementsByID.set(insertedElement.elementID.toKey(), insertedElement);
        const insertOperation = new InsertOperation(insertedElement.elementID, previousElement.elementID, nextElement.elementID, character);
        this.insertLog.set(insertedElement.elementID.toKey(), insertOperation);

        this.characterCounter++;


        return id
    }

    deleteElement(index) {
        const element = this.findVisibleElement(index);

        element.isDeleted = true;
        this.characterCounter--;
        const deleteOperation = new DeleteOperation(element.elementID);
        this.deleteLog.set(element.elementID.toKey(), deleteOperation);

        return element.elementID;
    }

    visibleContent() {
        const content = [];

        let current = this.start.right;

        while (current !== this.end) {
            if (!current.isDeleted) {
                content.push(current.content);
            }
            current = current.right;
        }

        return content.join("");
    }

    conflictZone(originID, rightID) {
        const origin = this.elementsByID.get(originID.toKey());
        const right = this.elementsByID.get(rightID.toKey());

        const conflictingElements = [];

        let current = origin.right;

        while (current !== right) {
            conflictingElements.push(current)
            current = current.right;
        }

        return conflictingElements;
    }

    isOriginAfter(insertOperation, conflictiveOperation) {
        let current = this.start.right;

        while (current !== this.end) {
            if (insertOperation === current) {
                return false
            }

            if (conflictiveOperation.origin === current) {
                return true
            }

            current = current.right
        }

        return false
    }

    findInsertionPoint(originID, rightID, newID) {
        const origin = this.elementsByID.get(originID.toKey());
        let left = origin
        let right = origin.right

        const conflictingElements = this.conflictZone(originID, rightID);

        for (const ops of conflictingElements) {
            if (this.isOriginAfter(origin, ops)) {
                break;
            }

            if (ops.elementID.clock < newID.clock) {
                left = ops;
                right = ops.right;
            } else if (ops.elementID.clock > newID.clock) {
                right = ops;
                break;
            } else {
                if (ops.elementID.client_id < newID.client_id) {
                    left = ops;
                    right = ops.right;
                } else {
                    right = ops;
                    break;
                }
            }
        }

        return [left, right];
    }

    processPending() {
        let progress = true;

        while (progress) {

            progress = false;

            for (const [key, value] of this.pendingInserts.entries()) {
                if (this.elementsByID.has(value.originID.toKey()) && this.elementsByID.has(value.rightID.toKey())) {
                    this.integrateInsert(value.newID, value.originID, value.rightID, value.content);
                    this.pendingInserts.delete(value.newID.toKey());
                    progress = true
                }
            }
        
        }
    }

    processPendingDeletes() {
        for (const [key, value] of this.pendingDeletes.entries()) {
            if (this.elementsByID.has(value.newID.toKey())) {
                this.integrateDeletion(value.newID);
                this.pendingDeletes.delete(value.newID.toKey());
            }
        }
    }

    integrateInsert(newID, originID, rightID, content) {
        
        let [left, right] = this.findInsertionPoint(originID, rightID, newID);
        const origin = this.elementsByID.get(originID.toKey());

        const element = new Element(newID, origin, left, right, content);

        left.right = element;
        right.left = element;

        this.elementsByID.set(newID.toKey(), element);
        const insertOperation = new InsertOperation(newID, originID, rightID, content);
        this.insertLog.set(newID.toKey(), insertOperation);

        this.characterCounter++;

    }

    remoteInsert(newID, originID, rightID, content) {
        if (this.elementsByID.has(newID.toKey())) {
            throw new Error("The value was already inserted.");
        }

        if (!this.elementsByID.has(originID.toKey()) || !this.elementsByID.has(rightID.toKey())) {
            this.pendingInserts.set(newID.toKey(), new PendingElement(newID, originID, rightID, content));
            const insertOperation = new InsertOperation(newID, originID, rightID, content);
            this.insertLog.set(newID.toKey(), insertOperation);
            throw new Error("Pending value.");
        }

        this.integrateInsert(newID, originID, rightID, content);

        this.processPending();

        this.processPendingDeletes(); 

    }

    integrateDeletion(elementID) {

        const element = this.elementsByID.get(elementID.toKey());
        element.isDeleted = true;

        this.characterCounter--;

        const deleteOperation = new DeleteOperation(element.elementID);
        this.deleteLog.set(element.elementID.toKey(), deleteOperation);
    }

    remoteDelete(elementID) {
        if (!this.elementsByID.has(elementID.toKey())) {
            this.pendingDeletes.set(elementID.toKey(), new PendingElement(elementID, null, null, '\x00'));
            const deleteOperation = new DeleteOperation(elementID);
            this.deleteLog.set(elementID.toKey(), deleteOperation);
            return
        }

        if (this.elementsByID.get(elementID.toKey()).isDeleted) {
            return
        }

        this.integrateDeletion(elementID);
    }
}